Build of OSS with 'yarn' broken by Electron 11 update

Building OSS from master on Windows started failing very recently. Bisecting has led me to this commit by @deepak1556  being the probable cause:

https://github.com/microsoft/vscode/commit/0baf6bb9cca978e6f3527b96331e27573b0dfb24

The commit title is `chore: update to electron 11 (#110759)`

When I checkout from here or later I get the problem. When I checkout its immediate predecessor commit (10d92e5efaf67a39660260f4fa5e44b3f847c597) the 'yarn' command completes successfully and I get a runnable OSS.

The 'yarn' failure when I build from the latest commit in master (d189ecf936) looks like this:
```
PS C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode> yarn
yarn install v1.22.5
$ node build/npm/preinstall.js
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@2.1.3: The platform "win32" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.1.1: The platform "win32" is incompatible with this module.
info "fsevents@2.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.7: The platform "win32" is incompatible with this module.
info "fsevents@1.2.7" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > eslint-plugin-mocha@8.0.0" has incorrect peer dependency "eslint@>=7.0.0".
warning " > gulp-tsb@4.0.5" has incorrect peer dependency "typescript@^3.0.0".
[4/4] Building fresh packages...
[6/19] ⢀ spdlog
[2/19] ⢀ native-is-elevated
[3/19] ⠠ native-keymap
[4/19] ⠠ native-watchdog
error C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap
Output:
C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
gyp info it worked if it ends with ok
gyp info using node-gyp@5.1.0
gyp info using node@12.18.4 | win32 | x64
gyp info find Python using Python version 2.7.15 found at "C:\Users\JohnM\.windows-build-tools\python27\python.exe"
gyp info find VS using VS2015 (14.0) found at:
gyp info find VS "C:\Program Files (x86)\Microsoft Visual Studio 14.0"
gyp info find VS run with --verbose for detailed information
gyp info spawn C:\Users\JohnM\.windows-build-tools\python27\python.exe
gyp info spawn args [
gyp info spawn args   'C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\gyp\\gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'msvs',
gyp info spawn args   '-I',
gyp info spawn args   'C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\vscode\\node_modules\\native-keymap\\build\\config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   'C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   'C:\\Users\\JohnM\\AppData\\Local\\node-gyp\\Cache\\11.0.2\\include\\node\\common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=C:\\Users\\JohnM\\AppData\\Local\\node-gyp\\Cache\\11.0.2',
gyp info spawn args   '-Dnode_gyp_dir=C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp',
gyp info spawn args   '-Dnode_lib_file=C:\\\\Users\\\\JohnM\\\\AppData\\\\Local\\\\node-gyp\\\\Cache\\\\11.0.2\\\\<(target_arch)\\\\node.lib',
gyp info spawn args   '-Dmodule_root_dir=C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\vscode\\node_modules\\native-keymap',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\vscode\\node_modules\\native-keymap\\build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn C:\Program Files (x86)\MSBuild\14.0\bin\MSBuild.exe
gyp info spawn args [
gyp info spawn args   'build/binding.sln',
gyp info spawn args   '/clp:Verbosity=minimal',
gyp info spawn args   '/nologo',
gyp info spawn args   '/p:Configuration=Release;Platform=x64'
gyp info spawn args ]
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
  string_conversion.cc
  keymapping.cc
  keyboard_win.cc
  win_delay_load_hook.cc
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(17): error C2144: syntax error: 'int' should be preceded by '}' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(17): error C2144: syntax error: 'int' should be preceded by ';' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(18): error C2065: 'kMayContainHeapPointers': undeclared identifier (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(18): error C2513: 'int': no variable declared before '=' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(18): error C2143: syntax error: missing ';' before ',' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(18): error C2059: syntax error: ',' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(22): error C2143: syntax error: missing ';' before '}' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(24): error C2059: syntax error: '}' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\cppgc/common.h(24): error C2143: syntax error: missing ';' before '}' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\v8-internal.h(16): error C2143: syntax error: missing ';' before '{' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\v8-internal.h(16): error C2447: '{': missing function header (old-style formal list?) (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
c:\users\johnm\appdata\local\node-gyp\cache\11.0.2\include\node\v8.h(1224): error C2039: 'Address': is not a member of 'v8::internal' (compiling source file ..\src\keymapping.cc) [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\vscode\node_modules\native-keymap\build\keymapping.vcxproj]
```
...

--- a/build/npm/preinstall.js
+++ b/build/npm/preinstall.js
@@ -28,7 +28,47 @@ if (!/yarn[\w-.]*\.js$|yarnpkg$/.test(process.env['npm_execpath'])) {
 	err = true;
 }
 
+const os = process.env['OS'];
+if (os.startsWith('Windows')) {
+	const vsVersion = getLatestSupportedVisualStudioVersion();
+	if (!vsVersion) {
+		console.error('\033[1;31m*** Please use Visual Studio 2017 or 2019.\033[0;0m');
+		err = true;
+	}
+}
+
 if (err) {
 	console.error('');
 	process.exit(1);
+}
+
+function getLatestSupportedVisualStudioVersion() {
+	const fs = require('fs');
+	const path = require('path');
+	// Translated over from
+	// https://source.chromium.org/chromium/chromium/src/+/master:build/vs_toolchain.py;l=140-175
+	const supportedVersions = ['2019', '2017'];
+
+	// VS installed in depot_tools for Googlers
+	if (process.env['DEPOT_TOOLS_WIN_TOOLCHAIN'] === '1') {
+		return supportedVersions[0];
+	}
+
+	// VS installed in system for everyone else
+	const availableVersions = [];
+	for (const version of supportedVersions) {
+		let vsPath = process.env[`vs${version}_install`];
+		if (vsPath && fs.existsSync(vsPath)) {
+			availableVersions.push(version);
+			break;
+		}
+		const programFiles86Path = process.env['ProgramFiles(x86)'];
+		vsPath = `${programFiles86Path}/Microsoft Visual Studio/${version}`;
+		const vsTypes = ['Enterprise', 'Professional', 'Community', 'Preview', 'BuildTools'];
+		if (programFiles86Path && vsTypes.some(vsType => fs.existsSync(path.join(vsPath, vsType)))) {
+			availableVersions.push(version);
+			break;
+		}
+	}
+	return availableVersions.length ? availableVersions[0] : undefined;
 }

--- a/build/npm/preinstall.js
+++ b/build/npm/preinstall.js
@@ -28,8 +28,7 @@ if (!/yarn[\w-.]*\.js$|yarnpkg$/.test(process.env['npm_execpath'])) {
 	err = true;
 }
 
-const os = process.platform;
-if (os === 'win32') {
+if (process.platform === 'win32') {
 	if (!hasSupportedVisualStudioVersion()) {
 		console.error('\033[1;31m*** Invalid C/C++ Compiler Toolchain. Please check https://github.com/microsoft/vscode/wiki/How-to-Contribute.\033[0;0m');
 		err = true;

--- a/build/npm/preinstall.js
+++ b/build/npm/preinstall.js
@@ -28,11 +28,10 @@ if (!/yarn[\w-.]*\.js$|yarnpkg$/.test(process.env['npm_execpath'])) {
 	err = true;
 }
 
-const os = process.env['OS'];
-if (os.startsWith('Windows')) {
-	const vsVersion = getLatestSupportedVisualStudioVersion();
-	if (!vsVersion) {
-		console.error('\033[1;31m*** Please use Visual Studio 2017 or 2019.\033[0;0m');
+const os = process.platform;
+if (os === 'win32') {
+	if (!hasSupportedVisualStudioVersion()) {
+		console.error('\033[1;31m*** Invalid C/C++ Compiler Toolchain. Please check https://github.com/microsoft/vscode/wiki/How-to-Contribute.\033[0;0m');
 		err = true;
 	}
 }
@@ -42,19 +41,13 @@ if (err) {
 	process.exit(1);
 }
 
-function getLatestSupportedVisualStudioVersion() {
+function hasSupportedVisualStudioVersion() {
 	const fs = require('fs');
 	const path = require('path');
 	// Translated over from
 	// https://source.chromium.org/chromium/chromium/src/+/master:build/vs_toolchain.py;l=140-175
 	const supportedVersions = ['2019', '2017'];
 
-	// VS installed in depot_tools for Googlers
-	if (process.env['DEPOT_TOOLS_WIN_TOOLCHAIN'] === '1') {
-		return supportedVersions[0];
-	}
-
-	// VS installed in system for everyone else
 	const availableVersions = [];
 	for (const version of supportedVersions) {
 		let vsPath = process.env[`vs${version}_install`];
@@ -63,12 +56,14 @@ function getLatestSupportedVisualStudioVersion() {
 			break;
 		}
 		const programFiles86Path = process.env['ProgramFiles(x86)'];
-		vsPath = `${programFiles86Path}/Microsoft Visual Studio/${version}`;
-		const vsTypes = ['Enterprise', 'Professional', 'Community', 'Preview', 'BuildTools'];
-		if (programFiles86Path && vsTypes.some(vsType => fs.existsSync(path.join(vsPath, vsType)))) {
-			availableVersions.push(version);
-			break;
+		if (programFiles86Path) {
+			vsPath = `${programFiles86Path}/Microsoft Visual Studio/${version}`;
+			const vsTypes = ['Enterprise', 'Professional', 'Community', 'Preview', 'BuildTools'];
+			if (vsTypes.some(vsType => fs.existsSync(path.join(vsPath, vsType)))) {
+				availableVersions.push(version);
+				break;
+			}
 		}
 	}
-	return availableVersions.length ? availableVersions[0] : undefined;
+	return availableVersions.length;
 }

